File: src/core/governance/ledger.ts
/**
 * The idempotency ledger — one flat JSON file recording what has already been
 * published, keyed by canonical URL.
 *
 * The key *is* the coordination. Whichever lane publishes a URL first writes
 * the key; every other lane — this extension, the MCP server, the Python CI
 * job — looks it up and skips. No lock, no lease, no shared process. That is
 * the whole design, and it only works if the two lanes agree on the file byte
 * for byte, which is why writes go through `pyJsonDump` with Python's
 * `indent=2, sort_keys=True, ensure_ascii=True` plus a trailing newline, and
 * why they are atomic (a half-written ledger is a double publish waiting to
 * happen).
 *
 * Keys beginning with `_` are metadata, not shares. `_meta` is the only one
 * this codebase writes today; the prefix convention means a future lane can
 * add its own without breaking anyone's iteration — provided everyone
 * enumerates shares through `shareEntries` rather than `Object.entries`.
 *
 * A missing ledger and a corrupt ledger both read as `{}`. "Nothing has been
 * published" is the safe answer to both: it makes the caller check its other
 * gates rather than trust a file it could not read.
 */

import { atomicWriteFile } from '../shared/atomic';
import { pyJsonDump, readJsonFile } from '../shared/jsonio';
import { utcStamp } from '../shared/timestamp';

/** The metadata key this codebase writes. Never enumerated as a share. */
export const TOKEN_KEY = '_meta';

export interface LedgerEntry {
  /** Stable identity of the published artifact, e.g. `jekyll:posts/x.md`. */
  urn: string;
  /** `utcStamp()` — second precision, no milliseconds. */
  posted_at: string;
  /** `article` or `update`. */
  type: string;
  /** Workspace-relative path of the draft or source file it came from. */
  source_file?: string;
  /** Id of the `PublishTarget` that produced the artifact. */
  target?: string;
}

/**
 * The file's shape. Entries are `Partial<LedgerEntry>` because the Python lane
 * and older versions of this one write keys we do not model — and dropping
 * them on a rewrite would be data loss.
 */
export type Ledger = Record<string, Partial<LedgerEntry> & Record<string, unknown>>;

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/** Read the ledger. Missing, unreadable and corrupt all yield `{}`. */
export async function loadLedger(ledgerPath: string): Promise<Ledger> {
  const raw = await readJsonFile<unknown>(ledgerPath);
  if (!isPlainObject(raw)) {
    return {};
  }
  const out: Ledger = {};
  for (const [key, value] of Object.entries(raw)) {
    if (isPlainObject(value)) {
      out[key] = value as Ledger[string];
    }
  }
  return out;
}

/**
 * Write the ledger. Exactly `json.dump(data, f, indent=2, sort_keys=True)` plus
 * a trailing newline: `sort_keys` makes the file order-independent so two lanes
 * never produce a reordering diff, and `ensure_ascii` escapes non-ASCII URLs
 * the same way on both sides.
 */
export async function saveLedger(ledgerPath: string, data: Ledger): Promise<void> {
  await atomicWriteFile(
    ledgerPath,
    `${pyJsonDump(data, { indent: 2, sortKeys: true, ensureAscii: true })}\n`,
  );
}

export async function getEntry(ledgerPath: string, url: string): Promise<Ledger[string] | undefined> {
  return (await loadLedger(ledgerPath))[url];
}

/** Has this canonical URL already been published, by either lane? */
export async function isPublished(ledgerPath: string, url: string): Promise<boolean> {
  const entry = await getEntry(ledgerPath, url);
  return typeof entry?.urn === 'string' && entry.urn !== '';
}

/**
 * Record a publish.
 *
 * The entry is **replaced**, not merged: a re-publish of the same URL is a new
 * fact about that URL, and merging would leave the ledger claiming a target or
 * a source file that no longer produced it. Absent options are omitted
 * entirely rather than written as `null`, because the Python lane's readers
 * test for key presence.
 */
export async function record(
  ledgerPath: string,
  url: string,
  urn: string,
  opts: { sourceFile?: string; kind?: string; target?: string } = {},
): Promise<LedgerEntry> {
  const data = await loadLedger(ledgerPath);
  const entry: LedgerEntry = {
    urn,
    posted_at: utcStamp(),
    type: opts.kind ?? 'article',
  };
  if (opts.sourceFile) {
    entry.source_file = opts.sourceFile;
  }
  if (opts.target) {
    entry.target = opts.target;
  }
  data[url] = { ...entry };
  await saveLedger(ledgerPath, data);
  return entry;
}

/**
 * The published shares, and only those: `_`-prefixed metadata keys are skipped,
 * and so is any entry without a `urn` (a half-written record, or a key another
 * lane parked there). **This is the only correct way to enumerate the ledger.**
 */
export function shareEntries(ledger: Ledger): Array<[string, LedgerEntry]> {
  const out: Array<[string, LedgerEntry]> = [];
  for (const [key, value] of Object.entries(ledger)) {
    if (key.startsWith('_')) {
      continue;
    }
    if (typeof value.urn !== 'string' || value.urn === '') {
      continue;
    }
    out.push([
      key,
      {
        urn: value.urn,
        posted_at: typeof value.posted_at === 'string' ? value.posted_at : '',
        type: typeof value.type === 'string' ? value.type : 'article',
        ...(typeof value.source_file === 'string' ? { source_file: value.source_file } : {}),
        ...(typeof value.target === 'string' ? { target: value.target } : {}),
      },
    ]);
  }
  return out;
}

/** Workspace-relative source files that have already been published. */
export function publishedSourceFiles(ledger: Ledger): Set<string> {
  const out = new Set<string>();
  for (const [, entry] of shareEntries(ledger)) {
    if (entry.source_file) {
      out.add(entry.source_file);
    }
  }
  return out;
}

/** Read the `_meta` block. Never a share; always an object. */
export async function readMeta(ledgerPath: string): Promise<Record<string, unknown>> {
  const meta = (await loadLedger(ledgerPath))[TOKEN_KEY];
  return meta === undefined ? {} : { ...meta };
}

/** Merge `patch` into the `_meta` block, leaving every share untouched. */
export async function writeMeta(ledgerPath: string, patch: Record<string, unknown>): Promise<void> {
  const data = await loadLedger(ledgerPath);
  const meta = { ...(data[TOKEN_KEY] ?? {}), ...patch };
  data[TOKEN_KEY] = meta as Ledger[string];
  await saveLedger(ledgerPath, data);
}

File: rails/lib/zer0_cms/distribution/ledger.rb
# frozen_string_literal: true

require "digest"
require "fileutils"
require "json"
require "securerandom"
require "tmpdir"

module Zer0Cms
  module Distribution
    # The idempotency ledger: one flat JSON file recording what has already
    # been published, keyed by what was published — an article's canonical URL,
    # or `draft:<path>` for a text update, which has no URL of its own.
    #
    # The key is the coordination. Whichever surface publishes a URL first
    # writes the key; every other surface looks it up and skips. That only
    # works if they agree on the file, so:
    #
    # * an entry carries the post URN under both `linkedin_urn` (the name the
    #   Python publisher wrote, and the one already in live ledgers) and `urn`
    #   (the name the VS Code extension reads) — a reader accepts either;
    # * the file is written as Python's `json.dump(indent=2, sort_keys=True)`
    #   plus a newline, byte-identical to what the other lanes write;
    # * a write is a read-modify-write under an exclusive lock, then an atomic
    #   rename, so two publishes on one machine cannot drop each other's entry.
    #
    # An attempt LinkedIn did not confirm is written down too, as an entry with
    # `state: unconfirmed` and no URN (`mark_unconfirmed`), because a post that
    # may exist must not be sent again by the next run.
    #
    # One rule differs from the extension on purpose. An unreadable ledger —
    # present but not JSON — is not read as "nothing published". For a
    # file-writing target that answer is safe; for LinkedIn it would re-post
    # every accepted draft. `readable?` is false, and the pipeline refuses.
    class Ledger
      META = "_meta"

      attr_reader :path

      def initialize(path)
        @path = path.to_s
      end

      def exist?
        File.file?(@path)
      end

      def readable?
        return true unless exist?

        JSON.parse(File.read(@path, encoding: "UTF-8")).is_a?(Hash)
      rescue JSON::ParserError, SystemCallError
        false
      end

      # The entries: every top-level value that is an object.
      def load
        load_raw.select { |_, value| value.is_a?(Hash) }
      end

      # The whole file as parsed, so a rewrite keeps keys of any shape it does
      # not own.
      def load_raw
        return {} unless exist?

        data = JSON.parse(File.read(@path, encoding: "UTF-8"))
        data.is_a?(Hash) ? data : {}
      rescue JSON::ParserError, SystemCallError
        {}
      end

      def entry(key)
        load[key.to_s]
      end

      def urn_for(key)
        self.class.urn_of(entry(key))
      end

      def published?(key)
        !urn_for(key).nil?
      end

      # The entry for an attempt LinkedIn did not confirm, or nil.
      def unconfirmed(key)
        entry = entry(key)
        entry if entry && self.class.urn_of(entry).nil? && entry["state"] == "unconfirmed"
      end

      # A create that ended in a 5xx, a dropped connection or a success with no
      # post id: the post may exist. The entry has no URN, so no reader counts
      # it as published, and the pipeline refuses the key until a person
      # records the post or forces past it. A recorded post is never demoted.
      def mark_unconfirmed(key, error:, kind:, author: nil, source_file: nil, now: Time.now)
        entry = { "state" => "unconfirmed", "attempted_at" => now.utc.strftime("%Y-%m-%dT%H:%M:%SZ"),
                  "error" => error.to_s[0, 300], "target" => "linkedin", "type" => kind }
        entry["author"] = author if author
        entry["source_file"] = source_file if source_file
        mutate { |data| data[key.to_s] = entry unless self.class.urn_of(data[key.to_s]) }
        entry
      end

      def self.urn_of(entry)
        return nil unless entry.is_a?(Hash)

        [entry["linkedin_urn"], entry["urn"]].map(&:to_s).find { |urn| !urn.empty? }
      end

      # The published shares, and only those: `_`-prefixed metadata and rows
      # without a URN are skipped. The only correct way to enumerate the file.
      def shares
        load.filter_map do |key, entry|
          next if key.start_with?("_")

          urn = self.class.urn_of(entry)
          urn ? [key, entry.merge("urn" => urn)] : nil
        end
      end

      # Replace the entry for `key`. A re-publish is a new fact about the key,
      # so nothing from the old entry is merged in.
      def record(key, urn:, kind:, author: nil, source_file: nil, image_urn: nil, now: Time.now)
        raise ArgumentError, "ledger key is empty" if key.to_s.empty? || key.to_s.start_with?("_")

        entry = { "linkedin_urn" => urn, "urn" => urn, "posted_at" => now.utc.strftime("%Y-%m-%dT%H:%M:%SZ"),
                  "target" => "linkedin", "type" => kind }
        entry["author"] = author if author
        entry["source_file"] = source_file if source_file
        entry["image_urn"] = image_urn if image_urn
        mutate { |data| data[key.to_s] = entry }
        entry
      end

      def meta
        load.fetch(META, {})
      end

      def write_meta(patch)
        mutate { |data| data[META] = data.fetch(META, {}).merge(patch) }
      end

      def self.serialize(data)
        "#{PyJson.dump(data, indent: 2, sort_keys: true, ensure_ascii: true)}\n"
      end

      private

      def mutate
        FileUtils.mkdir_p(File.dirname(@path))
        File.open(lock_path, File::RDWR | File::CREAT, 0o600) do |lock|
          lock.flock(File::LOCK_EX)
          raise ConfigError, "#{@path} is not valid JSON — refusing to rewrite it" unless readable?

          data = load_raw
          yield data
          write_atomically(self.class.serialize(data))
        end
      end

      # The lock lives outside the repository so it never shows up in git status.
      def lock_path
        File.join(Dir.tmpdir, "zer0-cms-ledger-#{Digest::SHA256.hexdigest(File.expand_path(@path))[0, 16]}.lock")
      end

      def write_atomically(text)
        dir = File.dirname(@path)
        temp = File.join(dir, ".#{File.basename(@path)}.#{Process.pid}.#{SecureRandom.hex(4)}.tmp")
        mode = exist? ? File.stat(@path).mode & 0o777 : 0o644
        File.open(temp, File::WRONLY | File::CREAT | File::EXCL, mode) do |file|
          file.write(text)
          file.flush
          file.fsync
        end
        File.rename(temp, @path)
      ensure
        FileUtils.rm_f(temp) if temp && File.exist?(temp)
      end
    end
  end
end
